### PR TITLE
Update NCF Model and README

### PR DESCRIPTION
**Pull request by AkhilByteWrangler**
### Changes

- Replaced the old NCF model with an improved version:
  - Added **user and item bias terms** to better capture global rating tendencies (e.g. some users rate higher than others on average).
  - Built a **deeper MLP architecture** with configurable hidden layers and **dropout** for regularization.
  - Training now runs for **10 epochs by default** instead of 100, reducing training time significantly.
  - Achieved **RMSE of ~0.55**, improved from previous ~6.xx, due to better architecture and training strategy.
  - Model now detects and uses **GPU (`cuda`) or MPS (Apple Silicon)** if available, falling back to CPU otherwise.

- Integrated **Optuna** for hyperparameter tuning:
  - Automatically searches over `embedding_dim`, `hidden_dims`, and learning rate.
  - Makes it easier to find the best configuration for new datasets or use cases.

- Improved model persistence:
  - Added logic to **save and load model checkpoints**.
  - Prevents retraining if a trained model is already available on disk.

- Updated the **README**:
  - Rewrote the Neural Collaborative Filtering section to reflect the new architecture, performance, and capabilities.

Best, 
Akhil (@AkhilByteWrangler)

### CODE REVIEW

**Comment by johnjoel2001** [comment]
Nice work on this update! 👏

The new NCF model looks much better, adding user and item bias is a smart move, and I like how you made the MLP deeper with dropout. It’s great that training is now faster with only 10 epochs and that you added support for GPU or MPS. 
Also, awesome to see the RMSE drop to ~0.55!

The Optuna integration is super helpful for tuning, and saving/loading models will save time. The README updates make everything easy to follow too.

Everything looks solid!!

**Review by haranku16 — CHANGES_REQUESTED** [review]
Thanks for improving the architecture and getting the performance where it needs to be! Could you just also update the evaluate script which uses the old class name?

File: scripts/ncf/model.py [hunk 1]
```diff
@@ -1,188 +1,265 @@
...
+        self.model.to(self.model.device)
 
     @staticmethod
-    def get_instance():
+    def _load():
         """
-        Loading an existing model if available; otherwise, train a new one.
+        Load model and encoders from disk
+        """
+        checkpoint = torch.load("models/enhanced_ncf.pth")
```
Comment (haranku16) [inline scripts/ncf/model.py:210]: Can you delete the old model pth and add the new one using git LFS?

**Review by haranku16 — APPROVED** [review]
